src/openfermion/ops/_interaction_tensor.py

```diff
@@ -84,103 +85,114 @@ def two_body_basis_change(two_body_tensor, rotation_matrix):
 
 
 class InteractionTensor(object):
-    """Class for storing data about the interactions between orbitals. Because
-    electrons interact pairwise, in second-quantization, all Hamiltonian terms
-    have either the form of a^\dagger_p a_q or a^\dagger_p a^\dagger_q a_r a_s.
-    The first of these terms is associated with the one-body Hamiltonian and
-    1-RDM and its information is stored in one_body_tensor. The second of these
-    terms is associated with the two-body Hamiltonian and 2-RDM and its
-    information is stored in two_body_tensor. Much of the functionality of this
-    class is redudant with FermionOperator but enables much more efficient
-    numerical computations in many cases, such as basis rotations.
+    """Class for storing tensor representations of operators that correspond
+    with multilinear polynomials in the fermionic ladder operators.
+    For instance, in a quadratic Hamiltonian (degree 2 polynomial),
+    there are only terms of the form a^\dagger_p a_q, and the coefficients
+    can be stored in an n_qubits x n_qubits matrix. Higher order terms would
+    be described with tensors of higher dimension. Note that each tensor must
+    have an even number of dimensions, since parity is conserved.
+    Much of the functionality of this class is redudant with FermionOperator
+    but enables much more efficient numerical computations in many cases,
+    such as basis rotations.
 
     Attributes:
-        n_qubits: The number of qubits on which the tensor
+        n_qubits(int): The number of qubits on which the tensor
             acts.
-        constant: A constant term in the operator given as a float.
-            For instance, the nuclear repulsion energy.
-        one_body_tensor: The coefficients of the 2D matrix terms. This is an
-            n_qubits x n_qubits numpy array of floats. For instance, the one
-            body term of MolecularOperator.
-        two_body_tensor: The coefficients of the 4D matrix terms. This is an
-            n_qubits x n_qubits x n_qubits x n_qubits numpy array offloats.
-            For instance, the two body term of MolecularOperator.
+        constant(complex): A constant term in the operator given as a complex
+            number. For instance, the nuclear repulsion energy.
+        n_body_tensors(dict): A dictionary storing the tensors describing
```

Comment:
I think the current implementation is already sufficiently general to support arbitrary marginals as subclasses. For instance, you could make a class called QuadraticOp, which stores quadratic operators, and in that case n_body_tensors[1] would be a 2n x 2n matrix, where n is the number of sites. Then you could say that the convention is that raising operators come before lowering operators, and so the (i, j + n) entry of this matrix stores the coefficient of a_i^\dagger a_j, and the (i + n, j) entry of this matrix stores the coefficient of a_i a_j^\dagger. You could even have subclasses where the coefficients correspond to other types of values; for instance, I am planning on creating a subclass called QuadraticMajoranaOp, where the tensors store coefficients of products of Majorana fermion operators, rather than the usual fermionic ladder operators.

Reply:
True, we could leverage the existing code to define different marginals but I think the object would be more powerful if a user was not required to modify code to expand functionality or the type of simulation they wanted to do.   The problem I have with your example is that if I want to define the hole and particle matrices with a 2n x 2n Tensor object I am wasting space with zeros.  If my Hamiltonian commutes with the number operator then a_{p}a_{q} terms and a_{p}^{\dag}a_{q}^{\dag} blocks are zero automatically.    Though, I am open to keeping the existing infrastructure because what I might be proposing could potentially break a lot of interfaces. @babbush for input.

Reply:
Actually, I think I understand what you mean that this could be used as a general object.  I could define an InteractionTensor for a 1-hole matrix corresponding to a_{p}a_{q}^{\dagger} with one matrix in it or an InteractionTensor for a 1-particle matrix corresponding to a_{p}^{\dagger}a_{q}.  Similarly I could do the same for any p-marginal.  It would be on the user to remember which InteractionTensor had which ladder operator ordering.   Is this what you had in mind?

Reply:
I might be misunderstanding but I think @kevinsung is actually proposing something different. He is suggesting that instead of an N by N matrix for n_body_tensors[1] there is a 2 N by 2 N matrix. Each dimension is doubled because there is now an entry for every possible raising and every possible lowering operator. For instance, if we have a 4 qubit systems then the matrix Kevin suggestions would be 8 by 8. Perhaps then a^\dagger_0 a_1 would be element (1, 2) and a_0 a^\dagger_1 would be element (0, 3).

This is a reasonable suggestion but I would vote against it. The reason is because the point of this class is essentially numerical efficiency and this is going to increase array size by 2^{k+1} where k is the many-body order.

I think I like Nick's first suggestion though. But Nick, is there ever a reason to store marginals other than the normal RDMs and the hole RDMs? For instance, could the key be (many_body_order, is_hole) where is_hole is 1 for hole RDMs and 1 for normal RDMs?

Reply:
@babbush The situation that you described, where n_body_tensors[1] is a 2N by 2N matrix, is indeed what I meant as a possible way to implement a general quadratic Hamiltonian. However, I intend for such functionality to be implemented in a subclass. Keep in mind that for the already implemented InteractionOperator, n_body_tensors[1] is an N by N matrix, where it is understood that there are only terms of the form a^\dagger_i a_j. That is, the rows are indexed by creation operators and the columns are indexed by lowering operators. I intend for this class to pretty much never be used directly, always appearing in practice in subclasses like InteractionOperator. What exactly the (i, j) entry of n_body_tensors[1] should be made clear in the subclass. For the case of InteractionOperator, the fact that it is particle-conserving means that n_body_tensors[1] only needs to be N by N, instead of 2N by 2N.

Reply:
To implement a general (not particle-number conserving) Hamiltonian as a subclass, I think n_body_tensors[1] would indeed have to be 2N by 2N. But if we use the Majorana basis, then it can be taken to be a matrix of real numbers. I intend to make such a subclass of this class (quadratic Hamiltonians in the Majorana basis).

Reply:
Kevin, I think you are suggesting that we can have a different derived class to represent things like hole matrices. Like, we'd have InteractionOperator always associated element (i, j) with a^\dagger_i a_j but we could have a different class inheriting from InteractionTensor that always associates (i, j) with a_i a^\dagger_j. But this requires that one implement a new subclass for each situation like this which might arise. I think Nick and I are pushing for having fewer derived classes which are more general.

For instance, Nick is suggesting the keys be tuples like (1, 0) to mean a^\dagger_p a_q matrix or (0, 1) to represent a_p a^\dagger_q matrix. I kind of like that idea. But this is likely to be confusing for people who only care about the normal one-body and two-body matrices. So perhaps we can implement Nick's suggestion but also associate the get/set methods automatically with (1, 0) and (1, 1, 0, 0) so people can "opt-in" to this more complicated functionality but don't necessarily need to worry about it. What do people think about that?